rfcs/20200504-tff-aggregators-placement.md

```diff
@@ -0,0 +1,226 @@
+for users of the lower-level Federated Core, at least until a higher level API is
+included that relies on it. Below, we detail how this change should be communicated
+by existing tutorials.
+
+### Tutorials and Examples
+Since this is a modification of an existing API, it likely does not warrant a new
+tutorial. We instead suggest modifying the existing
+[Part 1 Federated Core tutorial (FC 1)](https://www.tensorflow.org/federated/tutorials/custom_federated_algorithms_1)
+to include one or more federated computations that operate on Aggregator-placed
+data. We also considered modifying part 2 of the FC tutorial, but decided against
+that due to its stated goals.
+
+Concretely, we recommend two modifications to the FC 1 tutorial:
+- In the "Placement" section, the discussion will need to include the
+`tff.AGGREGATORS` placement. This section should stress that the placement can be
+considered optional, whereas the others (`tff.CLIENTS`, `tff.SERVER`) are strictly
+necessary for most interesting federated computations.
+- In the "Composing Federated Computations" section, we recommend adding a short
+sub-section or paragraph that describes how one might refactor the
+`get_average_temperature` function to perform its `federated_mean` with a placement
+of `tff.AGGREGATORS`. We include short and long form examples below for
+consideration.
+
+```python
+# short form
+@tff.federated_computation(tff.FederatedType(tf.float32, tff.CLIENTS))
+def get_average_temperature(sensor_readings):
+    averaged_temp = tff.federated_mean(sensor_readings, placement=tff.AGGREGATORS)
+    return tff.federated_collect(averaged_temp, placement=tff.SERVER)
+
+# long form
+@tff.federated_computation(tff.FederatedType(tf.float32, tff.CLIENTS))
+def get_average_temperature(sensor_readings):
+    collected_readings = tff.federated_collect(sensor_readings, placement=tff.AGGREGATORS)
+    num_clients = len(collected_readings)
+    total_temp = tff.sequence_sum(collected_readings)
+    return tff.federated_map(lambda x: x / num_clients, total_temp)
+```
+
+### Compatibility
+
+Since this design adds new functionality, it would change the public API. While TFF
+is still pre-1.0, it does not yet explicitly guarantee backwards compatibility of
+its public API. Nevertheless, we can hope to limit impact on the public API through
+judicious use of default keyword arguments.
+
+Concretely, we can maintain backwards compatibility for federated computations that
+gain a `placement` keyword argument by defaulting that argument to `tff.SERVER`. We
+recommend _judicious_ use because there may be instances where a change in semantic
+justifies a breaking change. These should be taken on a case-by-case basis, and we
+hope to clearly define and justify any breaking changes that might arise.
+
+This design does not significantly impact compatibility with the rest of the TF
+ecosystem.
+
+## Questions and Discussion Topics
+
```

Comment:
Perhaps you can add a pointer to [here](https://github.com/tensorflow/federated/blob/v0.13.1/tensorflow_federated/python/core/utils/encoding_utils.py#L306), which is an example of current inefficiency that would be possible to address using `tff.AGGREGATORS`.